### PR TITLE
Chapter 4 done

### PR DESCRIPTION
**Pull request by urmaul**
### Solutions for Chapter 4

cc @vrom911 @chshersh

Are you still doing reviews?

### CODE REVIEW

File: src/Chapter4.hs [hunk 1]
```diff
@@ -472,10 +487,12 @@ Implement the Applicative instance for our 'Secret' data type from before.
...
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+    pure = Reward
 
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    (<*>) (Reward f) (Reward a) = Reward (f a)
+    (<*>) (Trap e) _ = Trap e
+    (<*>) _ (Trap e) = Trap e
```
Comment (vrom911) [inline src/Chapter4.hs:495]: Nice! 👌🏼 
You could also reuse the `Functor` instance of `Secret` here and you would have only 2 cases 🙂

File: src/Chapter4.hs [hunk 2]
```diff
@@ -489,6 +506,14 @@ Implement the 'Applicative' instance for our 'List' type.
   type.
 -}
 
+instance Applicative List where
+    pure :: a -> List a
+    pure a = Cons a Empty
+
+    (<*>) :: List (a -> b) -> List a -> List b
+    (<*>) (Cons f ft) (Cons a at) = Cons (f a) (ft <*> at)
+    (<*>) _ _ = Empty
```
Comment (vrom911) [inline src/Chapter4.hs:515]: Great catch 👏🏼

File: src/Chapter4.hs [hunk 3]
```diff
@@ -629,7 +665,11 @@ Can you implement a monad version of AND, polymorphic over any monad?
 🕯 HINT: Use "(>>=)", "pure" and anonymous function
 -}
 andM :: (Monad m) => m Bool -> m Bool -> m Bool
-andM = error "andM: Not implemented!"
+andM a b = a >>= (andX b)
```
Comment (vrom911) [inline src/Chapter4.hs:668]: Nice lazy implementation 💯 
You don't need parentheses:
```suggestion
andM a b = a >>= andX b
```

File: src/Chapter4.hs [hunk 4]
```diff
@@ -673,6 +713,25 @@ Specifically,
...
+reverseTree :: Tree a -> Tree a
+reverseTree (Node val l r) = Node val (reverseTree r) (reverseTree l)
+reverseTree Leaf = Leaf
+
+
+treeToList :: Tree a -> List a
+treeToList (Node val l r) = flatten (list3 (treeToList l) (pure val) (treeToList r))
+  where
+    list3 :: a -> a -> a -> List a
+    list3 a b c = Cons a (Cons b (Cons c Empty))
```
Comment (vrom911) [inline src/Chapter4.hs:732]: Interesting!
I guess you could make it a bit more efficient if you implement append function for List instead. `append` also could be reused for `flatten` implementation.

File: src/Chapter4.hs [hunk 5]
```diff
@@ -673,6 +713,25 @@ Specifically,
  ❃ Implement the function to convert Tree to list
 -}
 
+data Tree a = Node a (Tree a) (Tree a) | Leaf deriving (Show)
```
Comment (vrom911) [inline src/Chapter4.hs:716]: Great Trees 🌴 🎄

**Review by vrom911 — APPROVED** [review]
:tada: Fantastic work! All CI is green 🍏 
Thank you so much for taking this course! I hope you had fun 🍰 👯 

We would really appreciate your [feedback on the course](https://docs.google.com/forms/d/e/1FAIpQLScBVhLxq5CgGnAfIGUE-fCoOUqeGkDY2HXzbT7KV2jjLOsmjQ/viewform)!

You can also ask any questions and we can discuss any topics you would like to dig into as well  :slightly_smiling_face:

**Comment by urmaul** [comment]
Thank you

**Pull request merged**